app/src/main/java/org/kiwix/kiwixmobile/wifi_hotspot/HotspotService.java

```diff
@@ -0,0 +1,178 @@
+    super.onCreate();
+
+    stopReceiver = new BroadcastReceiver() {
+      @Override
+      public void onReceive(Context context, Intent intent) {
+        if (intent != null && intent.getAction().equals(ACTION_STOP)) {
+          stopHotspotAndDismissNotification();
+        }
+      }
+    };
+
+    registerReceiver(stopReceiver, new IntentFilter(ACTION_STOP));
+  }
+
+  @Override public int onStartCommand(@NonNull Intent intent, int flags, int startId) {
+    switch (intent.getAction()) {
+
+      case ACTION_TOGGLE_HOTSPOT:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          if (hotspotManager.isHotspotStarted()) {
+            stopHotspotAndDismissNotification();
+          } else {
+            zimHostCallbacks.requestLocationAccess();
+          }
+        }
+        break;
+
+      case ACTION_LOCATION_ACCESS_GRANTED:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          hotspotManager.turnOnHotspot();
+        }
+        break;
+
+      case ACTION_START_SERVER:
+        if (webServerHelper.startServerHelper(
+            intent.getStringArrayListExtra(SELECTED_ZIM_PATHS_KEY))) {
+          zimHostCallbacks.onServerStarted(ServerUtils.getSocketAddress());
+          if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+            startForegroundNotificationHelper();
+          }
+          Toast.makeText(this, R.string.server_started__successfully_toast_message,
+              Toast.LENGTH_SHORT).show();
+        } else {
+          zimHostCallbacks.onServerFailedToStart();
+          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            stopForeground(true);
+            stopSelf();
+            hotspotNotificationManager.dismissNotification();
+          }
+        }
+
+        break;
+
+      case ACTION_STOP_SERVER:
+        stopHotspotAndDismissNotification();
+        break;
+      default:
+        break;
+    }
+    return START_NOT_STICKY;
+  }
+
+  @Nullable @Override public IBinder onBind(@Nullable Intent intent) {
+    return serviceBinder;
+  }
+
+  //Dismiss notification and turn off hotspot for devices>=O
+  private void stopHotspotAndDismissNotification() {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      hotspotManager.turnOffHotspot();
+    } else {
+      webServerHelper.stopAndroidWebServer();
+      zimHostCallbacks.onServerStopped();
+      stopForeground(true);
+      stopSelf();
+      hotspotNotificationManager.dismissNotification();
+    }
+  }
+
+  @Override
+  public void onDestroy() {
+    if (stopReceiver != null) {
+      unregisterReceiver(stopReceiver);
+    }
+    super.onDestroy();
+  }
+
+  public void registerCallBack(@Nullable ZimHostCallbacks myCallback) {
+    zimHostCallbacks = myCallback;
+  }
+
+  private void startForegroundNotificationHelper() {
+    startForeground(HOTSPOT_NOTIFICATION_ID,
+        hotspotNotificationManager.buildForegroundNotification(
+            getString(R.string.hotspot_running)));
```

Comment:
I don't think you need to pass this string in, just have it in HotspotNotificationManager

Reply:
Fixed